### ISSUE TITLE
[$250] Attempting to pay an employee with no deposit account requires the admin to click "Pay with Expensify" twice, resulting in duplicated confirmation messages

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#47503]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.20-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4706616
**Email or phone of affected tester (no customers):** qaooxx+57@gmail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:

```
Use Regions bank
Username: user_good | Password: pass_good
Account type: Plaid Saving (2nd option)
CompanyName: Alberta Bobbeth Charleson
CompanyTaxID: 123456789
First Name: Alberta
Last Name: Charleson
Last SSN numbers: 3333
```

### Setup - NewDot

1. Sign-up for a new account on expensify.com using a Gmail account > choose 1-9 in the qualifiers > get redirected to NewDot
2. Choose the "Manage my team's expenses" path in the onboarding modal to create a workspace on completion
3. Go to Settings > Workspaces > Click the workspace > Workflows > Enable approvals
4. Click "Connect bank account" to add a fully verified VBBA using the details above
5. Navigate to the Members tab of the workspace settings
7. Invite a new user as an employee (must be Gmail)
8. Verify the user workspace chat is displayed in the LHN

### Steps:
1. As Employee - Submit expenses in workspace chat
2. As Admin - Go to the expense report, approve it
3. Click Pay with Expensify (observe button still stays on the screen)
4. Click Pay with Expensify again (button disappears after second click)

## Expected Result:
Expense payment should be initiated after a single click of the `Pay with Expensify` button. Confirmation messages should only be displayed once for both admin and employee.

*Note that payment is on hold until the employee adds a bank account.

## Actual Result:
Expense payment is initiated only the `Pay with Expensify` button is clicked twice. Confirmation messages are displayed twice for both admin and employee. 

*Note that payment is on hold until the employee adds a bank account.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/82e046cb-7699-455a-ab20-17f3e9478de7

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011007936f05f04497</li>
        <li>Upwork Job ID: 1826365707781713597</li>
        <li>Last Price Increase: 2024-12-18</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

**Comment by greg-schroeder** [comment]
Sending through triage

**Comment by NJ-2020** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Expenses - Double Click requested & messages on Expensify Payment to No BA Employee
### What is the root cause of that problem?
When we click the approve button we invoke the `approveMoneyRequest` function, and the `approveMoneyRequest` function set the optimistic data the `stateNum` property to 3 which meaning the expense is approved without waiting until the API completely finish
https://github.com/Expensify/App/blob/42bed64c6a6d36bca317812d2a570a850f5ecc86/src/libs/actions/IOU.ts#L6918-L6931

And the Pay button briefly show because the expense is approved and the current isWaitingOnBankAccount is false
https://github.com/Expensify/App/blob/42bed64c6a6d36bca317812d2a570a850f5ecc86/src/libs/actions/IOU.ts#L6855-L6865

But after the `approveMoneyRequest` API finished, the BE response the `isWaitingOnBankAccount` property to true because i think we need to wait until the employee add the Bank Account, and if we wait a little bit of time we can see that the Pay button will disappear

### What changes do you think we should make in order to solve the problem?
I know that this is not as the Expected Result which we need still to display the Pay button
> One click wot Pay with Expensive should be enough. Confirmation messages under Admin and Employee should be displayed only once too

But i think we can't display the pay button because the BE returning the `isWaitingOnBankAccount` to true because we need to wait until the employee add the Bank account unless we remove the checking for `isWaitingOnBankAccount`

But here's my solution that will resolved the Pay button briefly appear and disappear
We can add new value inside the successData variable inside the `approveMoneyRequest` function that tell that the approved is successfully completed
https://github.com/Expensify/App/blob/42bed64c6a6d36bca317812d2a570a850f5ecc86/src/libs/actions/IOU.ts#L6948C11-L6967

### What alternative solutions did you explore? (Optional)
We can remove the `stateNum` inside the optimisticData

**Comment by NJ-2020** [comment]
I know my above proposal will not show the pay button because the BE returning the isWaitingOnBankAccount to true because we need to wait until the employee add the Bank account unless we remove the checking for isWaitingOnBankAccount

cc: @akinwale @greg-schroeder

**Comment by greg-schroeder** [comment]
Thanks for the extra context @NJ-2020 - next up is @akinwale to review

**Comment by greg-schroeder** [comment]
Go away Melvin

**Comment by trjExpensify** [comment]
Have a feeling this might have something to do with approvals in #wave-control perhaps. CC'ing @marcochavezf for vis.

**Comment by NJ-2020** [comment]
@akinwale Bump ^

**Comment by greg-schroeder** [comment]
Bump @akinwale

**Comment by akinwale** [comment]
@NJ-2020's [proposal ](https://github.com/Expensify/App/issues/47503#issuecomment-2308203265) can work here.

@NJ-2020 I believe the expected behaviour is that the Pay button should not be displayed after clicking it the first time while waiting for the response from the BE. The internal engineer can confirm this.

🎀👀🎀 C+ reviewed.

**Comment by srikarparsi** [comment]
Hey @marcochavezf, do you think you could take a look at this whenever you have a chance as it has to do with approval flows? I'm not too sure what we want to do here. I think best case is the pay button shows immediately and works optimistically if that's possible?

**Comment by greg-schroeder** [comment]
Awaiting review from Marco as a next step!

**Comment by marcochavezf** [comment]
Hey, apologies for the delay here, yeah I think we should show the pay button optimistically if the frontend already has the conditions to show it

**Comment by srikarparsi** [comment]
@akinwale do you think you could re-evaluate based on that whenever you get a chance?

**Comment by greg-schroeder** [comment]
bump @akinwale on the above

**Comment by akinwale** [comment]
@NJ-2020 Could you record a video of your solution so we can evaluate what the behaviour looks like? Thanks.

**Comment by NJ-2020** [comment]
@akinwale After looking into it, seems like BE issue, the first request the BE response `isWaitingOnBankAccount` to false which will keep the pay button still shows, but when trying again to pay for the second request the BE return the `isWaitingOnBankAccount` to true and the pay button disappear

**Comment by greg-schroeder** [comment]
Okay should I move this `Internal` then?

**Comment by srikarparsi** [comment]
Yes, I think that makes sense, changing this to internal. I don't currently have the bandwidth to take this one on so un-assigning myself. Do you think you could check if this fits into a wave/vip @greg-schroeder? I think that's in this [SO](https://stackoverflowteams.com/c/expensify/questions/18578)

**Comment by greg-schroeder** [comment]
Yeah I think it's wave-collect for sure

**Comment by greg-schroeder** [comment]
Still awaiting an internal volunteer here via `Hot Pick`

**Comment by trjExpensify** [comment]
@greg-schroeder can we clean up the title and the OP so it's coherent and makes sense? For example, these:

>Expenses - Double Click requested & messages on Expensify Payment to No BA Employee
>One click wot Pay with Expensive should be enough. Confirmation messages under Admin and Employee should be displayed only once too

**Comment by greg-schroeder** [comment]
Hey @trjExpensify oops, sorry! I'll do that now

**Comment by trjExpensify** [comment]
Thanks, that would be great.

**Comment by greg-schroeder** [comment]
Fixed, it should be much more readable now

**Comment by trjExpensify** [comment]
Cool, thanks! I don't think they need to do that dance on OldDot anymore given all collect workspaces are created with isPolicyExpenseChatEnabled and they can add a VBBA on NewDot. So I've modified that to streamline the steps.

**Comment by greg-schroeder** [comment]
Ah, good point! Thanks Tom

**Comment by greg-schroeder** [comment]
This is a Hot Pick, awaiting a volunteer

**Comment by trjExpensify** [comment]
@techievivek I'm pretty sure you're looking into this in another issue? 

Either way, moving to the #quality GH project in the wake of wave-collect being closed.

**Comment by greg-schroeder** [comment]
@techievivek is what Tom said [here](https://github.com/Expensify/App/issues/47503#issuecomment-2413546487) the case?

**Comment by greg-schroeder** [comment]
Bump @techievivek

**Comment by techievivek** [comment]
Yeah, it seems like an issue with creating duplicated action when trying to pay someone. I can handle this.

**Comment by greg-schroeder** [comment]
Whatcha thinking as next steps for this one @techievivek?

**Comment by techievivek** [comment]
Bumped the thread where I started the discussion https://expensify.slack.com/archives/C03TQ48KC/p1731421494485509

**Comment by greg-schroeder** [comment]
Thanks Vivek!

**Comment by techievivek** [comment]
We have an agreed solution from the Slack thread https://expensify.slack.com/archives/C03TQ48KC/p1731421494485509

Frontend:
We would like to hide the button in the frontend if the `expensify_waitingOnEmployeeBankAccount` rNVP is present on the report. IIRC, we used to have this in the frontend code but can't see it now.

Backend:
We need to throw if we are re-attempting to pay a request where the above rNVP is already added.

I think we can work on these changes parallel. Going to send this to external so we can find the contributor for the frontend part of the change. I can work on the backend changes for it. Thanks.

**Comment by akinwale** [comment]
> We have an agreed solution from the Slack thread https://expensify.slack.com/archives/C03TQ48KC/p1731421494485509
> 
> Frontend: We would like to hide the button in the frontend if the `expensify_waitingOnEmployeeBankAccount` rNVP is present on the report. IIRC, we used to have this in the frontend code but can't see it now.
> 
> Backend: We need to throw if we are re-attempting to pay a request where the above rNVP is already added.
> 
> I think we can work on these changes parallel. Going to send this to external so we can find the contributor for the frontend part of the change. I can work on the backend changes for it. Thanks.

@NJ-2020 If you're still interested, could you update your proposal based on this? Thanks.

**Comment by techievivek** [comment]
Not overdue, waiting for the proposals.

**Comment by techievivek** [comment]
Bumped in slack for proposals https://expensify.slack.com/archives/C01GTK53T8Q/p1733208086766289

**Comment by greg-schroeder** [comment]
Still awaiting outcome of Slack conversation

**Comment by greg-schroeder** [comment]
Same

**Comment by ijmalik** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue.
When attempting to pay an employee without a deposit account, the admin is required to click "Pay with Expensify" twice, leading to duplicate confirmation messages.

## What is the root cause of that problem?

https://github.com/Expensify/App/blob/7925c9b555af60156c62947ae63b517d6c31587f/src/libs/actions/IOU.ts#L7103

### Step-1
When we press ```Approve``` button, ```Pay with Expensify```  button immediately appear optimistically without waiting [API-Response-Approve](https://github.com/user-attachments/files/18057460/47503-API-Response-Approve.json).
In this response "**isWaitingOnBankAccount**": **false**. So ```Pay with Expensify``` button stay their due to above condition

### Step-2
When we **first time** press ```Pay with Expensify```  button, **Payment complete** message shown and ```Pay with Expensify```  button hidden optimistically.

In the [API PayMoneyRequest Response ](https://github.com/user-attachments/files/18057834/47503-API-Response-PayMoneyRequest-First.json) "**isWaitingOnBankAccount**": **false**. So ```Pay with Expensify``` button come back and shown again due to same above condition

### Step-3
When we **second time** press ```Pay with Expensify```  button, **Payment complete** message again shown and ```Pay with Expensify```  button hidden optimistically.
In the [Second API PayMoneyRequest Response](https://github.com/user-attachments/files/18057959/47503-API-Response-PayMoneyRequest-Second.json) "**isWaitingOnBankAccount**": **true**. So ```Pay with Expensify``` button remains hidden due to same above condition.

So root cause is that in Step-2 response has "**isWaitingOnBankAccount**": **false**

## What changes do you think we should make in order to solve the problem?

In Step-2 server response with "**isWaitingOnBankAccount**" having its value **true**

## What alternative solutions did you explore? (Optional)

In the  [First API PayMoneyRequest Response ](https://github.com/user-attachments/files/18057834/47503-API-Response-PayMoneyRequest-First.json) it has action "REIMBURSEMENTQUEUED". If payment is already queued then ```Pay with Expensify``` button should not be shown 


**Replace**
https://github.com/Expensify/App/blob/7925c9b555af60156c62947ae63b517d6c31587f/src/libs/actions/IOU.ts#L7098-L7110

**With**


```
    const isPayAtEndExpenseReport = ReportUtils.isPayAtEndExpenseReport(iouReport?.reportID, transactions);
    const reimbursementQueuedActions = getReimbursementQueuedActions(iouReport?.reportID);
    const hasReimbursementQueuedActions = reimbursementQueuedActions.length > 0;
    return (
        isPayer &&
        !isOpenExpenseReport &&
        !iouSettled &&
        !iouReport?.isWaitingOnBankAccount &&
        !hasReimbursementQueuedActions &&
        reimbursableSpend !== 0 &&
        !isChatReportArchived &&
        !isAutoReimbursable &&
        !shouldBeApproved &&
        !isPayAtEndExpenseReport
    );
}

/**
 * Retrieves all reimbursement queued actions for a given IOU report ID.
 *
 * @param {string} iouReportID - The ID of the IOU report.
 * @returns {Array<ReportAction<typeof CONST.REPORT.ACTIONS.TYPE.REIMBURSEMENT_QUEUED>>} - List of reimbursement queued actions.
 */
function getReimbursementQueuedActions(
    iouReportID: string
): Array<ReportAction<typeof CONST.REPORT.ACTIONS.TYPE.REIMBURSEMENT_QUEUED>> {
    // Retrieve the report actions for the given report ID, defaulting to an empty object if none exist.
    const reportActions = allReportActions?.[`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${iouReportID}`] ?? {};

    // Filter and return only the reimbursement queued actions.
    return Object.values(reportActions).filter((reportAction): reportAction is ReportAction<typeof CONST.REPORT.ACTIONS.TYPE.REIMBURSEMENT_QUEUED> => 
        ReportActionsUtils.isReimbursementQueuedAction(reportAction)
    );
}
```


**Before Change:**
https://drive.google.com/file/d/1c3MxrBpKqG0SMqRQHagKfQ5LpBAbGIX5/view?usp=sharing

**After Change (alternative solutions)**

https://github.com/user-attachments/assets/556e1b9a-1fe6-4e92-9172-d871c1f3f562

**Comment by greg-schroeder** [comment]
@akinwale next up to the review the proposal above!

**Comment by akinwale** [comment]
@ijmalik Can you confirm that your proposed solution works based on this [comment](https://github.com/Expensify/App/issues/47503#issuecomment-2527634903)?

@techievivek Is the backend implementation complete and ready for testing?

**Comment by ijmalik** [comment]
> @ijmalik Can you confirm that your proposed solution would work based on this [comment](https://github.com/Expensify/App/issues/47503#issuecomment-2527634903)?
> 
Hi @akinwale,

Yes, I confirm that my proposed [solution](https://github.com/Expensify/App/issues/47503#issuecomment-2527634903) is expected to work as intended. I have tested it thoroughly in my local environment to ensure its accuracy and functionality.

Please feel free to reach out if you have any additional questions or concerns.

**Comment by techievivek** [comment]
Not overdue, I will review this tomorrow.

**Comment by greg-schroeder** [comment]
Thanks @techievivek

**Comment by techievivek** [comment]
Looks like this already got fixed somewhere else? @ijmalik are you able to reproduce this now?

**Comment by techievivek** [comment]
Yeah, just tested with both pay with Expensify and pay elsewhere and both worked without any duplication.

**Comment by techievivek** [comment]
Going to ask QA to confirm if they can reproduce it or not.

**Comment by techievivek** [comment]
Requested for re-test here: https://expensify.slack.com/archives/C9YU7BX5M/p1734512368639129

**Comment by ijmalik** [comment]
> Looks like this already got fixed somewhere else? @ijmalik are you able to reproduce this now?

Hi @techievivek,
Thank you for checking. I am still able to reproduce the issue. Please find the attached video for your reference.

https://github.com/user-attachments/assets/d11efa9a-a5f8-49f9-9e83-4679fc440356

**Comment by ijmalik** [comment]
> Yeah, just tested with both pay with Expensify and pay elsewhere and both worked without any duplication.

Hi @techievivek,

Could you please confirm if the prerequisite actions were performed before testing?

```
Use Regions bank
Username: user_good | Password: pass_good
Account type: Plaid Saving (2nd option)
CompanyName: Alberta Bobbeth Charleson
CompanyTaxID: 123456789
First Name: Alberta
Last Name: Charleson
Last SSN numbers: 3333
```

## Setup - NewDot
```
Sign-up for a new account on expensify.com using a Gmail account > choose 1-9 in the qualifiers > get redirected to NewDot
Choose the "Manage my team's expenses" path in the onboarding modal to create a workspace on completion
Go to Settings > Workspaces > Click the workspace > Workflows > Enable approvals
Click "Connect bank account" to add a fully verified VBBA using the details above
Navigate to the Members tab of the workspace settings
Invite a new user as an employee (must be Gmail)
Verify the user workspace chat is displayed in the LHN
```

**Comment by izarutskaya** [comment]
Still reproducible for tester

https://github.com/user-attachments/assets/037425ec-7272-4e07-9dc8-823e21c07abe

**Comment by techievivek** [comment]
Not overdue, pushing the backend fix today.

**Comment by techievivek** [comment]
Pushed the backend changes for this, we will not need any frontend changes afterwards.

**Comment by ijmalik** [comment]
Dear @techievivek, @akinwale, and @greg-schroeder,

I hope this message finds you well.

I’m reaching out to follow up on the proposal I submitted 21 days ago, which has been under review for the past 5 days. I completely understand that schedules can be quite busy, especially around the holidays, but I wanted to kindly inquire if there are any updates on the review process.

Please let me know if there’s anything further I can provide or assist with to help move things along.

Thank you so much for your time and support—I truly appreciate it!

Looking forward to hearing from you.

**Comment by techievivek** [comment]
Hey @ijmalik, I investigated this last week and realized that it was a bug on the backend where we were not correctly setting the `isWaitingOnBankAccount` status for the expense report in the first `PayMoneyRequest` call. So I have pushed a fix on the backend for this, so we will not need any additional changes for it in the frontend for now. The changes will be live on staging in a day or two so we can probably test it further and confirm if the bug is fixed fully.

**Comment by techievivek** [comment]
Backend PR is in review with Vit. Hopefully, it should be merged/deployed by this week. Once that's live we can ask QA team to confirm the bug is fixed.

**Comment by greg-schroeder** [comment]
Thanks @techievivek!

**Comment by greg-schroeder** [comment]
Deployed to prod. I'll ask QA to retest

**Comment by kavimuru** [comment]
Not reproducible anymore.

https://github.com/user-attachments/assets/d54bde49-7f05-445a-98ea-3ae755fda382

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by techievivek** [comment]
Closing this since it's no longer reproducible.

**Comment by ijmalik** [comment]
[akinwale](https://github.com/akinwale)
> > [@ijmalik](https://github.com/ijmalik) Can you confirm that your proposed solution would work based on this [comment](https://github.com/Expensify/App/issues/47503#issuecomment-2527634903)?
>
> Hi [@akinwale](https://github.com/akinwale),
>
> Yes, I confirm that my proposed [solution](https://github.com/Expensify/App/issues/47503#issuecomment-2527634903) is expected to work as intended. I have tested it thoroughly in my local environment to ensure its accuracy and functionality.



**Dear @JmillsExpensify, @garrettmknighton**

This particular issue had remained unresolved for several months, and I invested significant time and effort to identify the root cause. Through detailed investigation, I discovered that the issue originated from the backend API response, which had previously gone undetected.
While implementing the solution itself was straightforward, the real challenge was the in-depth analysis and expertise required to diagnose the problem accurately. Resolving this long-standing issue has contributed to the overall stability and functionality of the system.
Given the time, skill, and dedication I devoted to addressing this matter, I would sincerely appreciate your consideration for appropriate compensation.
Thank you for your understanding and support.

**Comment by greg-schroeder** [comment]
Hey @ijmalik! Sorry for missing your message before. I wanted to make sure your request was given its due attention. 👍

I know it's probably not the answer you're looking for, but this isn't a situation where we'd offer compensation. I'll point you towards our guidelines on contributing [here](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions):

>Please be aware that compensation for any support in solving an issue is provided entirely at Expensify’s discretion. Personal time or resources applied towards investigating a proposal will not guarantee compensation. Compensation is only guaranteed to those who [propose a solution and get hired for that job](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job). We understand there may be cases where a selected proposal may take inspiration from a previous proposal. Unfortunately, it’s not possible for us to evaluate every individual case and we have no process that can efficiently do so. Issues with higher rewards come with higher risk factors so try to keep things civil and make the best proposal you can. Once again, any information provided may not necessarily lead to you getting hired for that issue or compensated in any way.

In this case, you weren't hired for the job, nor was your proposal selected for implementation at any point. Also, based on the timeline I can see, I can't say an egregiously unreasonable amount of time passed between your proposal and the root cause was identified and the issue was fixed elsewhere. It can be frustrating when it works out this way, but it happens -- it's just sort of the nature of a super fast-paced open source project with many hundreds of contributors. 

With that said, we definitely recognize and value the skillset required to investigate and make a proposal of the caliber you created. However, this just isn't a situation that would lead to a payment given the circumstances. Despite that, we'd still love to see you participate as part of the community as you obviously have a lot to offer!
